agent/proxycfg/snapshot.go

```diff
@@ -482,6 +482,10 @@ func (c *ConfigSnapshot) MeshGatewayValidExportedServices() []structs.ServiceNam
 			continue // not possible
 		}
 
+		if _, ok := c.MeshGateway.ServiceGroups[svc]; !ok {
```

Comment:
I'm a little unfamiliar with the ServiceGroups. So it looks like the `MeshGatewayValidExportedServices` function is used in a few places. Is this going to cause issues elsewhere in the code, or have you already verified all of that logic?

Reply:
I verified changes for the exported services in peering and and wan federated clusters.

Based on my understanding, `MeshGatewayValidExportedServices` is used to return a list of services that are exported. However, since the exported service could be deregistered, these services should be removed from the mesh gateway, i.e., removing from the `ServiceGroups`. The current code doesn't remove them because deregister service doesn't trigger the event at https://github.com/hashicorp/consul/blob/985a4ee1b1644f9ab74de90d39dfc0a4233d8d60/agent/proxycfg/mesh_gateway.go#L695  ;
Actually, the notify is unwatched, so it will never triggered
https://github.com/hashicorp/consul/blob/985a4ee1b1644f9ab74de90d39dfc0a4233d8d60/agent/proxycfg/mesh_gateway.go#L232-L262

This is noted https://github.com/hashicorp/consul/blob/985a4ee1b1644f9ab74de90d39dfc0a4233d8d60/agent/proxycfg/mesh_gateway.go#L256-L260

Reply:
Thanks for taking the time to explain all of that.